flake.nix

```
@@ -0,0 +1,42 @@
+{
+  description = "Rust devshell for Iced";
+
+  inputs = {
+    nixpkgs.url      = "github:NixOS/nixpkgs/nixos-unstable";
+    rust-overlay.url = "github:oxalica/rust-overlay";
+    flake-utils.url  = "github:numtide/flake-utils";
+  };
+
+  outputs = { nixpkgs, rust-overlay, flake-utils, ... }:
+    flake-utils.lib.eachDefaultSystem (system:
+      let
+        overlays = [ (import rust-overlay) ];
+        pkgs = import nixpkgs {
+          inherit system overlays;
+        };
+      in
+      with pkgs;
+      {
+        devShells.default = mkShell rec {
+          nativeBuildInputs = [
+            bacon
+            cargo-udeps
+            cargo-edit
+            rust-analyzer
+            cargo-deny
+            rust-bin.stable.latest.default
+
+            pkg-config
+            dbus
+          ];
+          buildInputs = [
+            libxkbcommon
+          ];
+          runtimeDependencies = [
+            wayland
+          ];
+          LD_LIBRARY_PATH = lib.makeLibraryPath (nativeBuildInputs ++ buildInputs ++ runtimeDependencies);
```

```suggestion
          runtimeDependencies = [
            wayland
          ];
```

This is what is happening by default in `mkShell`, we don't have to be explicit. Also, the name of this environment variable vary, depending on the platform.

---

For some reason, if I remove this, I get hit with an error:
```
thread 'main' panicked at /home/seb/projects/iced/winit/src/lib.rs:84:10:
Create event loop: Os(OsError { line: 81, file: "/home/seb/.cargo/git/checkouts/winit-865c6d5afb3ec9c3/11414b6/src/platform_impl/linux/wayland/event_loop/mod.rs", error: WaylandError(Connection(NoWaylandLib)) })
```

---

Yeah, this is because you use the `runtimeDependencies` which is an attribute to add libraries to rpath but not into `LD_LIBRARY_PATH`. Initially you've done it manually via `lib.makeLibraryPath` but this can be avoided altogether.

---

So, I've been trying different things, but maybe I'm misunderstanding something.
I moved all the dependencies to `buildInputs` and also removed the whole `LD_LIBRARY_PATH = lib.makeLibraryPath (nativeBuildInputs);` line (see the updated file). I still get the same runtime error when running `cargo run --package tour`.

Do you have any suggestions on what I could be missing?

---

>     2. The above flake does not account for the Rust toolchain (for my use-case I have Rust installed on my system for now)
I think it would be better to include it, because the idea of dev flakes is to have a working environment out of the box.

>     3. This assumes a Vulkan backend rather than something like OpenGL.
This doesn't take into account old systems that do not support Vulkan. Since Iced is using wgpu, which is essentially an abstraction layer on top of Vulkan, OpenGL, DirectX, Metal, it would be better to not include this info.

> I'm moreso curious if this flake structure is good practice and will generally work for environments like `iced`.

So for flakes, as far as I understand, we need to be more generic. It has to work for every system supported by Iced (and by Nix, of course). But I'm far from being an expert. I still don't grasp the best practices with Nix.